### DISCUSSION TITLE
Setup https for running code-server from ssh remote server

### DISCUSSION THREAD

**Discussion by duckhaitong** [coder/code-server#4045] [Q&A]
Hi everyone, I'm tryna setup https for my code-server, which is run on my ssh remote server. I tried to use --link but it did not work. Because I can only expose code-server in local browser when HTTP listening on 0.0.0.0 in remote server while --link always force to use 127.0.0.1. self-sign certificate did not work for me as well, but I might be wrong. I've heard that proxy can help, but I don't understand it. Any suggest for me also be appreciated. Thanks in advance.

**Comment by lg-nineoneninethreenine** [comment]
I am unsure about this but in the Coder YouTube channel, they demonstrated how `--link` works. Note that thie flag is in beta so there are errors. Mainly, the latency. I have WSL, a superfast network but the flag doesn't work. Make sure that the latency is below 75 ms, because I think the latency is the problem. If you need more help, I say that @jsjoeio has the better answer for you. Moreover, put a screenshot of the error here so that we can understand.
Note: If you use Google Cloud, use the Cloud Console and try. Try it from the server's page, be it Google, Azure, Oracle or anything.

**Comment by duckhaitong** [comment]
Thanks @im-coder-lg for your suggest. Here is the log when running with --link flag 
```
[2021-08-26T11:25:51.299Z] info  code-server 3.11.1 c680aae973d83583e4a73dc0c422f44021f0140e
[2021-08-26T11:25:51.300Z] info  Using user-data-dir ~/.local/share/code-server
[2021-08-26T11:25:51.315Z] info  Using config file ~/.config/code-server/config.yaml
[2021-08-26T11:25:51.315Z] info  HTTP server listening on http://127.0.0.1:19277 (randomized by --link)
[2021-08-26T11:25:51.315Z] info    - Authentication is disabled (disabled by --link)
[2021-08-26T11:25:51.315Z] info    - Not serving HTTPS (disabled by --link)
[2021-08-26T11:25:51.351Z] info  ping server: dial server: failed to WebSocket dial: failed to send handshake request: Get "https://cloud.coder.com/latency": Proxy Authentication Required
[2021-08-26T11:25:51.354Z] error --link agent exited with 1
```

I tried generating self-signed certificate from OpenSSL (use with --cert --cert-key, import in my local machine) but the connection is still insecure. I think it's because my certificate does not have "Allows secure communication on the Internet" (I check it via MMC console). So I'm now stuck in self-signed certificate approach. Could "proxy-approach" do the trick ?

**Reply by lg-nineoneninethreenine** [reply]
Proxy auth... Have you tried Let's Encrypt certs?

**Reply by duckhaitong** [reply]
I can't use Port forwarding, Let's Encrypt, due to limited privileges, can not use sudo command.

**Reply by duckhaitong** [reply]
by the way, I'm running code-server on 0.0.0.0:8080. with port >= 1024 it does not require root permission.

**Reply by lg-nineoneninethreenine** [reply]
Wait what? You said that you run it on a server but you can't use the sudo command?

**Reply by duckhaitong** [reply]
Yes, simply download .tar.gz and decompress it

**Reply by jsjoeio** [reply]
Hmm...you can't port forward so ngrok or localtunnel wouldn't work, right?

Maybe @code-asher has a suggestion

**Reply by duckhaitong** [reply]
@jsjoeio I think so. I've not used ngrok or localtunnel so far.

**Reply by duckhaitong** [reply]
@code-asher Could you help me please .

**Reply by code-asher** [reply]
Link dials out to our cloud but it looks like for some reason that connection is failing.

`Proxy Authentication Required` makes it look like you have a proxy between yourself and the internet that requires authentication?

I am not sure how to give link the authentication information. It is possible that we need to modify link to support this use case.

**Reply by duckhaitong** [reply]
@code-asher Yes, I have a proxy server between the local and the internet. Code-server hasn't supported this sort of case yet?

**Reply by code-asher** [reply]
code-server does but I think link does not. I created an issue here: https://github.com/cdr/link/issues/1

**EDIT**: Looks like the link repository is currently private but I think we will make it public again once the rewrite is complete?
